### PR TITLE
h2olog: make -t to accept a glob pattern like "quicly:*"

### PR DESCRIPTION
**Pull request by gfx**
As discussed at https://github.com/h2o/h2o/pull/2492#issuecomment-730737191.

This PR has slightly changed the semantics of `-l` in order to test what`-t` actually select, like:

```
h2olog quic -t 'quicly:*' -l
```

The behavior of `-l` does not change without `-t`.

### CODE REVIEW

File: src/h2olog/h2olog.cc [hunk 1]
```diff
@@ -224,6 +226,19 @@ static void lost_cb(void *context, uint64_t lost)
     tracer->handle_lost(lost);
 }
 
+static size_t add_matched_usdts(std::vector<h2o_tracer::usdt> &selected_usdts, const std::vector<h2o_tracer::usdt> &available_usdts,
+                                const char *pattern)
+{
+    size_t added = 0;
+    for (auto usdt : available_usdts) {
```
Comment (syohex) [inline src/h2olog/h2olog.cc:233]:
```suggestion
    for (const auto& usdt : available_usdts) {
```

You can avoid copying loop variable by using reference.

Comment (gfx) [reply src/h2olog/h2olog.cc]: Nice catch. Thanks.

File: src/h2olog/h2olog.cc [hunk 2]
```diff
@@ -272,7 +285,7 @@ int main(int argc, char **argv)
             debug++;
             break;
         case 'l':
-            for (const auto &usdt : available_usdts) {
+            for (const auto &usdt : selected_usdts.empty() ? available_usdts : selected_usdts) {
```
Comment (kazuho) [inline src/h2olog/h2olog.cc:288]: I'm not sure if we want to incorporate this change in its current form.

The first problem is that the definition of `-l` changes depending on how `-t` is used. Moreover, the behavior depends on the order those two options are being used. Consider the case where the user specifies `-l -t quicly.*`. Wouldn't the output be surprising?

I'm not sure if we need this, considering the fact that we report an error when the glob does not match anything.

**Review by kazuho — COMMENTED** [review]
LGTM modulo the point below.

File: src/h2olog/h2olog.cc (hunk 2 above)
Comment (gfx) [reply src/h2olog/h2olog.cc]: That's rather a feature for h2olog developers and I have no objection to drop it. done at 5f57b8a379cbcb38ab99decb74c4f14ecb684551

**Review by kazuho — APPROVED** [review]
Thank you for the changes. LGTM. BTW, I like the use of fnmatch :+1:

File: src/h2olog/h2olog.cc [hunk 3]
```diff
@@ -46,7 +47,8 @@ Optional arguments:
     -h Print this help and exit
     -l Print the list of available tracepoints and exit
     -s RESPONSE_HEADER_NAME A response header name to show, e.g. "content-type"
-    -t TRACEPOINT A tracepoint, or fully-qualified probe name, to show, e.g. "quicly:accept"
+    -t TRACEPOINT A tracepoint, or fully-qualified probe name, to show,
+                  accepting a glob pattern, e.g. "quicly:accept", "h2o:*"
```
Comment (janaiyengar) [inline src/h2olog/h2olog.cc:51]:
```suggestion
                  including a glob pattern, e.g. "quicly:accept", "h2o:*"
```

**Review by janaiyengar — APPROVED** [review]
super minor suggestion. Thanks for doing this though -- this is going to be super useful!

**Pull request merged**